### PR TITLE
Add wait_for_slot_event and get_slot_event

### PR DESCRIPTION
**Pull request by arjennienhuis**

### CODE REVIEW

**Review by ionut-arm — APPROVED** [review]
👍🏻 Thanks for the patch, it's unfortunate that SoftHSM doesn't support this... ohwell

File: cryptoki/src/context/slot_token_management.rs [hunk 1]
```diff
@@ -162,4 +168,28 @@ impl Pkcs11 {
...
+        unsafe {
+            let mut slot: CK_SLOT_ID = 0;
+            let wait_for_slot_event = get_pkcs11!(self, C_WaitForSlotEvent);
+            let rv = wait_for_slot_event(flags, &mut slot, std::ptr::null_mut());
+            Rv::from(rv).into_result()?;
+            Ok(Slot::new(slot))
+        }
+    }
+
+    /// wait for slot events (insertion or removal of a token)
```
Comment (wiktor-k) [inline cryptoki/src/context/slot_token_management.rs:182]:
```suggestion
    /// Wait for slot events (insertion or removal of a token)
```

File: cryptoki/src/context/slot_token_management.rs [hunk 2]
```diff
@@ -162,4 +168,28 @@ impl Pkcs11 {
...
+    pub fn wait_for_slot_event(&self) -> Result<Slot> {
+        self.wait_for_slot_event_impl(0)
+    }
+
+    /// get the latest slot event (insertion or removal of a token)
```
Comment (wiktor-k) [inline cryptoki/src/context/slot_token_management.rs:187]:
```suggestion
    /// Get the latest slot event (insertion or removal of a token)
```

File: cryptoki/tests/basic.rs [hunk 3]
```diff
@@ -1003,3 +1003,28 @@ fn aes_gcm_with_aad() -> TestResult {
...
     Ok(())
 }
+
+#[test]
+#[serial]
+fn get_slot_event() -> TestResult {
+    // Not implemented in SoftHSMv2
+    // https://github.com/opendnssec/SoftHSMv2/issues/370
+    let (pkcs11, _slot) = init_pins();
+    let event = pkcs11.get_slot_event().unwrap();
```
Comment (wiktor-k) [inline cryptoki/tests/basic.rs:1013]: I think `?` should work here as expected (due to [`testresult`](https://crates.io/crates/testresult)):

```suggestion
    let event = pkcs11.get_slot_event()?;
```

File: cryptoki/tests/basic.rs [hunk 4]
```diff
@@ -1003,3 +1003,28 @@ fn aes_gcm_with_aad() -> TestResult {
...
+#[test]
+#[serial]
+fn wait_for_slot_event() -> TestResult {
+    // Not implemented in SoftHSMv2
+    // https://github.com/opendnssec/SoftHSMv2/issues/370
+    let (pkcs11, _slot) = init_pins();
+    let res = pkcs11.wait_for_slot_event();
+
+    match res {
+        Ok(_) | Err(Error::Pkcs11(RvError::FunctionNotSupported)) => Ok(()),
```
Comment (wiktor-k) [inline cryptoki/tests/basic.rs:1027]: :thinking: I think this entire `match` can be shortened to something like this:

```rust
assert!(matches!(res, Err(Error::Pkcs11(RvError::FunctionNotSupported))));
```

**Review by wiktor-k — APPROVED** [review]
This looks very nice, found only tiny things. Thanks for your contribution! :pray:

File: cryptoki/src/context/slot_token_management.rs (hunk 1 above)
Comment (arjennienhuis) [reply cryptoki/src/context/slot_token_management.rs]: Done.

File: cryptoki/tests/basic.rs (hunk 3 above)
Comment (arjennienhuis) [reply cryptoki/tests/basic.rs]: Done.

File: cryptoki/tests/basic.rs (hunk 4 above)
Comment (arjennienhuis) [reply cryptoki/tests/basic.rs]: I still included the error message. `assert_matches!(...)` would do that automatically but is "nightly-only".

Comment (wiktor-k) [reply cryptoki/tests/basic.rs]: Yep, got it! 👍 Thanks!

**Comment by wiktor-k** [comment]
Aaaand it's merged! Thanks for your time going through our suggestions and... have a nice day! 😊

**Pull request merged**